File: src/lib/format.ts
import type { Period } from '../db/schema';

const pad2 = (n: number) => String(n).padStart(2, '0');

export function moneyParts(n: number, currency = 'BRL', locale = 'pt-BR'): { symbol: string; value: string } {
  const parts = new Intl.NumberFormat(locale, { style: 'currency', currency }).formatToParts(n);
  let symbol = '';
  let value = '';
  for (const p of parts) {
    if (p.type === 'currency' || p.type === 'minusSign' || p.type === 'plusSign') symbol += p.value;
    else if (p.type !== 'literal') value += p.value;
  }
  return { symbol, value };
}

export function money(n: number, currency = 'BRL', locale = 'pt-BR'): string {
  const { symbol, value } = moneyParts(n, currency, locale);
  return `${symbol} ${value}`;
}

export function fmtDate(d?: string | null, locale = 'pt-BR'): string {
  if (!d) return '—';
  return new Date(d + 'T00:00:00').toLocaleDateString(locale, {
    day: '2-digit',
    month: 'short',
  });
}

export function toISO(d: unknown): string | null {
  if (!d) return null;
  if (typeof d === 'string') return d.slice(0, 10);
  if (d instanceof Date) return `${d.getFullYear()}-${pad2(d.getMonth() + 1)}-${pad2(d.getDate())}`;
  return null;
}

export function periodForDate(
  date?: string | null,
  tripStartDate?: string | null,
): Period | null {
  if (!date || !tripStartDate) return null;
  return date < tripStartDate ? 'BEFORE' : 'DURING';
}

export function dateRange(start: string, end: string): string[] {
  const out: string[] = [];
  const e = new Date(end + 'T00:00:00');
  for (let d = new Date(start + 'T00:00:00'); d <= e; d.setDate(d.getDate() + 1)) {
    out.push(`${d.getFullYear()}-${pad2(d.getMonth() + 1)}-${pad2(d.getDate())}`);
  }
  return out;
}

function isNextDay(a: string, b: string): boolean {
  const d = new Date(a + 'T00:00:00');
  d.setDate(d.getDate() + 1);
  return `${d.getFullYear()}-${pad2(d.getMonth() + 1)}-${pad2(d.getDate())}` === b;
}

export interface CityBlock {
  city: string;
  start: string;
  end: string;
  days: string[];
}

export function groupCityBlocks(
  days: string[],
  cities: Record<string, string | null | undefined>,
): CityBlock[] {
  const blocks: CityBlock[] = [];
  for (const d of days) {
    const city = cities[d];
    if (!city) continue;
    const last = blocks[blocks.length - 1];
    if (last && last.city === city && isNextDay(last.end, d)) {
      last.end = d;
      last.days.push(d);
    } else {
      blocks.push({ city, start: d, end: d, days: [d] });
    }
  }
  return blocks;
}

File: lib/format.dart
import 'package:intl/intl.dart';
import 'package:centavoo/models/transaction.dart' show periodBefore, periodDuring;

String appLocale = 'pt_BR';

String money(double n, {String currency = 'BRL', String? locale}) {
  final loc = locale ?? appLocale;
  final symbol = currency == 'BRL' ? 'R\$' : currency;
  final isNegative = n < 0;
  final valueStr = NumberFormat.currency(locale: loc, symbol: '', decimalDigits: 2).format(n.abs()).trim();
  return '${isNegative ? '-' : ''}$symbol $valueStr';
}

double? parseAmountInput(String raw) {
  var s = raw.trim();
  if (s.isEmpty) return null;
  final hasDot = s.contains('.');
  final hasComma = s.contains(',');
  if (hasDot && hasComma) {
    final lastDot = s.lastIndexOf('.');
    final lastComma = s.lastIndexOf(',');
    s = lastComma > lastDot ? s.replaceAll('.', '').replaceAll(',', '.') : s.replaceAll(',', '');
  } else if (hasComma) {
    s = s.replaceAll(',', '.');
  }
  return double.tryParse(s);
}

String fmtDate(String? d, {String? locale}) {
  if (d == null) return '—';
  final loc = locale ?? appLocale;
  final date = DateTime.parse(d);
  final pattern = loc.startsWith('en') ? 'MMM d' : "dd 'de' MMM";
  return DateFormat(pattern, loc).format(date);
}

String? periodForDate(String? date, String? tripStartDate) {
  if (date == null || tripStartDate == null) return null;
  return date.compareTo(tripStartDate) < 0 ? periodBefore : periodDuring;
}

String isoDate(DateTime d) =>
    '${d.year.toString().padLeft(4, '0')}-${d.month.toString().padLeft(2, '0')}-${d.day.toString().padLeft(2, '0')}';

List<String> dateRange(String start, String end) {
  final out = <String>[];
  final endDate = DateTime.parse('${end}T00:00:00');
  for (var d = DateTime.parse('${start}T00:00:00'); !d.isAfter(endDate); d = d.add(const Duration(days: 1))) {
    out.add(isoDate(d));
  }
  return out;
}

class CityBlock {
  final String city;
  final String start;
  final String end;
  final List<String> days;

  CityBlock({required this.city, required this.start, required this.end, required this.days});
}

bool _isNextDay(String a, String b) {
  final d = DateTime.parse('${a}T00:00:00').add(const Duration(days: 1));
  return isoDate(d) == b;
}

List<CityBlock> groupCityBlocks(List<String> days, Map<String, String> cities) {
  final blocks = <CityBlock>[];
  for (final d in days) {
    final city = cities[d];
    if (city == null || city.isEmpty) continue;
    final last = blocks.isNotEmpty ? blocks.last : null;
    if (last != null && last.city == city && _isNextDay(last.end, d)) {
      blocks[blocks.length - 1] = CityBlock(city: city, start: last.start, end: d, days: [...last.days, d]);
    } else {
      blocks.add(CityBlock(city: city, start: d, end: d, days: [d]));
    }
  }
  return blocks;
}
